internal/vulnstore/postgres/getupdateoperations.go

```diff
@@ -180,50 +256,45 @@ func getUpdateOperations(ctx context.Context, pool *pgxpool.Pool, updater ...str
 		rows.Close()
 	}
 
-	// Take care to close the rows object on every iteration.
-	var rows pgx.Rows
-	for _, u := range updater {
+	var q string
+	var label string
+	switch kind {
+	case "":
+		q = query
+		label = "query"
+	case driver.EnrichmentKind:
+		q = queryEnrichment
+		label = "query_enrichment"
+	case driver.VulnerabilityKind:
+		q = queryVulnerability
+		label = "query_vulnerability"
+	}
 
-		start := time.Now()
+	start := time.Now()
+	rows, err := tx.Query(ctx, q, updater)
+	switch {
+	case err == nil:
+	case errors.Is(err, pgx.ErrNoRows):
+		return nil, nil
+	default:
+		return nil, fmt.Errorf("failed to get distinct updates: %w", err)
+	}
+	getUpdateOperationsCounter.WithLabelValues(label).Add(1)
+	getUpdateOperationsDuration.WithLabelValues(label).Observe(time.Since(start).Seconds())
 
-		rows, err = tx.Query(ctx, query, u)
-		switch {
-		case err == nil:
-		case errors.Is(err, pgx.ErrNoRows):
-			zlog.Warn(ctx).Str("updater", u).Msg("no update operations for this updater")
-			rows.Close()
-			continue
-		default:
+	for rows.Next() {
+		var uo driver.UpdateOperation
+		err := rows.Scan(
+			&uo.Ref,
+			&uo.Updater,
+			&uo.Fingerprint,
+			&uo.Date,
+		)
+		if err != nil {
 			rows.Close()
-			return nil, fmt.Errorf("failed to retrieve update operation for updater %v: %w", updater, err)
-		}
-		ops := []driver.UpdateOperation{}
-
-		getUpdateOperationsCounter.WithLabelValues("query").Add(1)
-		getUpdateOperationsDuration.WithLabelValues("query").Observe(time.Since(start).Seconds())
-
-		for rows.Next() {
-			ops = append(ops, driver.UpdateOperation{})
-			uo := &ops[len(ops)-1]
-			err := rows.Scan(
-				&uo.Ref,
-				&uo.Updater,
-				&uo.Fingerprint,
-				&uo.Date,
-			)
-			if err != nil {
-				rows.Close()
-				return nil, fmt.Errorf("failed to scan update operation for updater %q: %w", u, err)
-			}
-		}
-		rows.Close()
-		if err := rows.Err(); err != nil {
-			return nil, err
+			return nil, fmt.Errorf("failed to scan update operation for updater %q: %w", uo.Updater, err)
 		}
-		out[u] = ops
```

Comment:
I'm a little worried about excess copying doing it this way, but it's certainly simpler to read.

Reply:
Where are the concerns?

Reply:
The previous version allocates a local slice, grows it and populates in-place, then adds it to the map, so the slices in the map never need to be updated once they escape. This version allocates a `uo` local, populates it, then potentially reallocates the updater's slice copies it into the slice.